main/src/ca/uwaterloo/flix/language/phase/EarlyTreeShaker.scala

```diff
@@ -0,0 +1,388 @@
+
+    case Expression.InvokeStaticMethod(_, args, _, _, _, _) =>
+      visitExps(args)
+
+    case Expression.GetField(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.PutField(_, exp1, exp2, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2)
+
+    case Expression.GetStaticField(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.PutStaticField(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.NewObject(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.NewChannel(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.GetChannel(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.PutChannel(exp1, exp2, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2)
+
+    case Expression.SelectChannel(rules, default, _, _, _, _) =>
+      visitExps(rules.map(_.chan)) ++ visitExps(rules.map(_.exp)) ++ default.map(visitExp).getOrElse(Set.empty)
+
+    case Expression.Spawn(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.Lazy(exp, _, _) =>
+      visitExp(exp)
+
+    case Expression.Force(exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.FixpointConstraintSet(_, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointLambda(_, _, _, _, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointMerge(_, _, _, _, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointSolve(_, _, _, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointFilter(_, _, _, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointInject(_, _, _, _, _, loc) =>
+      throw InternalCompilerException(s"Unexpected expression near: ${loc.format}.")
+
+    case Expression.FixpointProject(_, exp, _, _, _, _) =>
+      visitExp(exp)
+
+    case Expression.Reify(_, _, _, _, _) =>
+      Set.empty
+
+    case Expression.ReifyType(_, _, _, _, _, _) =>
+      Set.empty
+
+    case Expression.ReifyEff(_, exp1, exp2, exp3, _, _, _, _) =>
+      visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
+
+    case Expression.Do(_, exps, _, _, _) =>
+      visitExps(exps)
+
+    case Expression.Resume(exp, _, _) =>
+      visitExp(exp)
+
+    case Expression.TryWith(exp, _, rules, _, _, _, _) =>
+      visitExp(exp) ++ visitExps(rules.map(_.exp))
+
+    case Expression.Without(exp, _, _, _, _, _) =>
+      visitExp(exp)
+
+  }
+
+  /**
+    * Returns the function symbols reachable from `es`.
+    */
+  private def visitExps(es: List[Expression]): Set[ReachableSym] = es.map(visitExp).fold(Set())(_ ++ _)
+
+
+  /**
+    * Represents a super type for reachable symbols in the AST.
+    */
+  sealed trait ReachableSym
```

Comment:
I would replace this by:

```
case class Reachable(defSyms: Set[...], sigSyms: [...], ...
```

with a `++` method.

This is both cleaner and faster 👍

Reply:
I'm not sure this works with `ParOps.parReachable`

Reply:
OK, that makes sense.